interop-reactiveStreams/shared/src/main/scala/scalaz/zio/interop/reactiveStreams/StreamPublisher.scala

```diff
@@ -18,35 +18,19 @@ class StreamPublisher[R, E <: Throwable, A](
           demand  <- Queue.unbounded[Long]
           _       <- Task(subscriber.onSubscribe(createSubscription(subscriber, demand)))
           control = Stream.fromQueue(demand).flatMap(n => Stream.unfold(n)(n => if (n > 0) Some(((), n - 1)) else None))
-          fiber   <- stream.toQueue().use(takesToCallbacks(subscriber, _, control, demand)).fork
+          fiber <- stream
+                    .zip(control)
```

Comment:
As you've noted before correctly, if `stream` has already failed but no demand has been issued, this will hang. So I think for now it'd be best to manually dequeue a demand element inside the call to `stream.foreach` rather than zip.

Reply:
I still need to unfold the demand. That would leave me with another fromQueue ... toQueue construct. Do you have something specific in mind, that you could provide as guidance? For now, I've switched to zipWith, which hides the complexity of toQueue, but fails early, as is uses toQueue internally.